### PR TITLE
sync-diff-inspector: add annotation for create table info

### LINKED ISSUE #678: sync diff cannot handle the table with int type primary key when clustered index is disabled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
set tidb_enable_clustered_index = off;
create table t (i int auto_increment primary key) shard_row_id_bits=2;
```

./bin/sync_diff_inspector -config=config.toml
```
[data-sources]
[data-sources.mysql1]
    host = "120.92.92.145"
    port = 4000
    user = "root"
    password = ""

[data-sources.tidb0]
    host = "120.92.92.145"
    port = 4000
    user = "root"
    password = ""

[task]
    output-dir = "./output"
```

2. What did you expect to see?
no error

3. What did you see instead?
```
github.com/pingcap/errors.AddStack
    /Users/yusp/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/errors.go:174
github.com/pingcap/errors.Trace
    /Users/yusp/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20211224045212-9687c2b0f87c/juju_adaptor.go:15
github.com/pingcap/tidb/ddl.BuildTableInfoWithStmt
    /Users/yusp/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20220918132100-29f83a0b2592/ddl/ddl_api.go:2228
github.com/pingcap/tidb/ddl.buildTableInfoWithCheck
    /Users/yusp/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20220918132100-29f83a0b2592/ddl/ddl_api.go:2149
github.com/pingcap/tidb/ddl.BuildTableInfoFromAST
    /Users/yusp/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20220918132100-29f83a0b2592/ddl/ddl_api.go:2143
github.com/pingcap/tidb-tools/pkg/dbutil.GetTableInfoBySQL
    /Users/yusp/work/tidb-tools/pkg/dbutil/table.go:59
github.com/pingcap/tidb-tools/pkg/dbutil.GetTableInfo
    /Users/yusp/work/tidb-tools/pkg/dbutil/table.go:47
github.com/pingcap/tidb-tools/sync_diff_inspector/source.initTables
    /Users/yusp/work/tidb-tools/sync_diff_inspector/source/source.go:323
github.com/pingcap/tidb-tools/sync_diff_inspector/source.NewSources
    /Users/yusp/work/tidb-tools/sync_diff_inspector/source/source.go:117
main.(*Diff).init
    /Users/yusp/work/tidb-tools/sync_diff_inspector/diff.go:139
main.NewDiff
    /Users/yusp/work/tidb-tools/sync_diff_inspector/diff.go:97
main.checkSyncState
    /Users/yusp/work/tidb-tools/sync_diff_inspector/main.go:117
main.main
    /Users/yusp/work/tidb-tools/sync_diff_inspector/main.go:104
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
v4.0.12

5. which tool are you using?

6. what version of tool are you using (`pump -V` or `tidb-lightning -V` or `syncer -V`)?
v4.0.12, v6.1.0

### PR DESCRIPTION
**Pull request by Leavrth**
Signed-off-by: Leavrth <jianjun.liao@outlook.com>

<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve?
<!--
Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.
-->

Issue Number: close #678 

### What is changed and how it works?

for version < 5, regard primary key as nonclustered.
so we can add annotation for create table info, if there is no annotation about clustered.

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test
 - Manual test (add detailed scripts or steps below)
 - No code

```
$ ./sync_diff_inspector --config config1.toml
debug: show table `db_test`.`t`'s create table info:
CREATE TABLE `t` (
  `i` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`i`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30001/*!90000 SHARD_ROW_ID_BITS=2 */
debug: show table `db_test`.`t0`'s create table info:
CREATE TABLE `t0` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */
debug: show table `db_test`.`t1`'s create table info:
CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */,
  KEY `term` (`term`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */
A total of 3 tables need to be compared


debug: show table `db_test`.`t`'s create table info:
CREATE TABLE `t` (
  `i` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`i`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30001/*!90000 SHARD_ROW_ID_BITS=
Progress [>------------------------------------------------------------] 0% 0/0
debug: show table `db_test`.`t1`'s create table info:
CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */,
  KEY `term` (`term`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */
debug: show table `db_test`.`t0`'s create table info:
CREATE TABLE `t0` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
Comparing the table structure of ``db_test`.`t`` ... equivalent
Comparing the table structure of ``db_test`.`t1`` ... equivalent
Comparing the table structure of ``db_test`.`t0`` ... equivalent
Comparing the table data of ``db_test`.`t1`` ... equivalent
Comparing the table data of ``db_test`.`t`` ... equivalent
Comparing the table data of ``db_test`.`t0`` ... equivalent
_____________________________________________________________________________
Progress [============================================================>] 100% 0/0
A total of 3 table have been compared and all are equal.
You can view the comparision details through '/tmp/sync-diff.output.2022-09-23T09.50.11+0800/sync_diff.log'
```

```SQL
mysql> show create table db_test.t;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                  |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `i` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`i`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30001/*!90000 SHARD_ROW_ID_BITS=2 */ |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)

mysql> show create table db_test.t0;
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                        |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t0    | CREATE TABLE `t0` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */ |
+-------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)                                                                                                                                                                                                         

mysql> show create table db_test.t1;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                               |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `term` (`term`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */ |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### CODE REVIEW

File: pkg/dbutil/common.go [hunk 1]
```diff
@@ -189,6 +190,42 @@ func CloseDB(db *sql.DB) error {
...
+//	+-------+--------------------------------------------------------------------+
+//	| Table | Create Table                                                                                                                              |
+//	+-------+--------------------------------------------------------------------+
+//	| itest | CREATE TABLE `itest` (
+//		`id` int(11) DEFAULT NULL,
+//		`name` varchar(24) DEFAULT NULL,
+//		PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
+//		) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+//	+-------+--------------------------------------------------------------------+
+func addClusteredAnnotationForPrimaryKey(raw string) (string, error) {
```
Comment (jackysp) [inline pkg/dbutil/common.go]: I think it should check if the tidb_version() is below v5, and the table has `PKisHandle` flag.
If version < 5 {
If `PKisHandle`,
add /*T![clustered_index] CLUSTERED */,
else
add /*T![clustered_index] NONCLUSTERED */.
}

**Comment by Leavrth** [comment]
1. table's pkishandle is true
```SQL
mysql> select _tidb_rowid from db_test.t1 limit 1;
ERROR 1054 (42S22): Unknown column '_tidb_rowid' in 'field list'
```

```
$ ./sync_diff_inspector --config config1.toml
debug: show table `db_test`.`t1`'s create table info:
CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
A total of 1 tables need to be compared


debug: show table `db_test`.`t1`'s create table info:
CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
Comparing the table structure of ``db_test`.`t1`` ... equivalent
Comparing the table data of ``db_test`.`t1`` ... equivalent
_____________________________________________________________________________
Progress [============================================================>] 100% 0/0
A total of 1 table have been compared and all are equal.
You can view the comparision details through '/tmp/sync-diff.output.2022-09-23T12.36.47+0800/sync_diff.log'
```

2. table's pkishandle is false
```SQL
mysql> select _tidb_rowid from db_test.t1 limit 0;
Empty set (0.00 sec)

mysql> select _tidb_rowid from db_test.t0 limit 0;
Empty set (0.00 sec)
```

```
$ ./sync_diff_inspector --config config1.toml
debug: show table `db_test`.`t0`'s create table info:
CREATE TABLE `t0` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */

debug: show table `db_test`.`t1`'s create table info:
CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin

A total of 2 tables need to be compared


debug: show table `db_test`.`t1`'s create table info:
CREATE TABLE `t1` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
Progress [>------------------------------------------------------------] 0% 0/0
debug: show table `db_test`.`t0`'s create table info:
CREATE TABLE `t0` (
  `id` bigint(20) NOT NULL,
  `term` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin/*!90000 SHARD_ROW_ID_BITS=2 */
Comparing the table structure of ``db_test`.`t1`` ... equivalent
Comparing the table structure of ``db_test`.`t0`` ... equivalent
Comparing the table data of ``db_test`.`t1`` ... equivalent
Comparing the table data of ``db_test`.`t0`` ... equivalent
_____________________________________________________________________________
Progress [============================================================>] 100% 0/0
A total of 2 table have been compared and all are equal.
You can view the comparision details through '/tmp/sync-diff.output.2022-09-23T12.41.47+0800/sync_diff.log'
```

**Review by jackysp — APPROVED** [review]
LGTM

File: pkg/dbutil/table.go [hunk 2]
```diff
@@ -27,12 +29,89 @@ import (
...
+	return false
+}
+
+func GetTableInfoWithVersion(ctx context.Context, db QueryExecutor, schemaName string, tableName string, version string) (*model.TableInfo, error) {
+	createTableSQL, err := GetCreateTableSQL(ctx, db, schemaName, tableName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if strings.Contains(version, "TiDB-v4") {
```
Comment (jackysp) [inline pkg/dbutil/table.go]: It looks OK, because v3 is the too old version.

File: pkg/dbutil/table.go [hunk 3]
```diff
@@ -27,12 +29,89 @@ import (
...
+			replaceString = AnnotationClusteredReplaceString
+		} else {
+			replaceString = AnnotationNonClusteredReplaceString
+		}
+		createTableSQL, err = addClusteredAnnotationForPrimaryKey(createTableSQL, replaceString)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	log.Info("123", zap.String("123", createTableSQL), zap.String("version", version))
```
Comment (3pointer) [inline pkg/dbutil/table.go]:
```suggestion
```

File: pkg/dbutil/table.go (hunk 2 above)
Comment (3pointer) [reply pkg/dbutil/table.go]: can we use `semver.Version`?

File: pkg/dbutil/table.go (hunk 3 above)
Comment (Leavrth) [reply pkg/dbutil/table.go]: :joy:

**Review by 3pointer — APPROVED** [review]

**Pull request merged**